### PR TITLE
[luci] Issue warnings when weights cannot be accurately represented

### PR DESCRIPTION
**Pull request by AShedko**
This PR logs warnings when weights cannot be accurately represented.
Another way to implement #8889

Signed-off-by: a.shedko <a.shedko@samsung.com>

### CODE REVIEW

**Comment by seanshpark** [comment]
> compiler/luci/pass/src/QuantizeBias.cpp:221:27: error: comparison of integer expressions of different signedness: ‘int32_t’ {aka ‘int’} and ‘std::vector<float>::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
03:10:22      for (int32_t i = 0; i < qp->min.size(); i++)
03:10:22                          ~~^~~~~~~~~~~~~~~~

File: compiler/luci/pass/src/QuantizationUtils.cpp [hunk 1]
```diff
@@ -395,4 +399,36 @@ void quant_const(luci::CircleConst *node, loco::DataType quant_type)
...
+      return 32;
+    case loco::DataType::S64:
+      return 64;
+    default:
+      return 64; // a safe large default
+  }
+}
+
+} // namespace
+bool range_check(float min, float max, loco::DataType dtype)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizationUtils.cpp:428]: please add empty line between these to lines

**Comment by seanshpark** [comment]
If the codes are necessary in release build, its OK, if not, it would be nice if only work in debug mode.
one way I can think of is to put it inside predefined function/macro only valid in debug mode like `assert()`.

**Comment by AShedko** [comment]
> If the codes are necessary in release build, its OK,
I would like warnings to appear in release mode, because we deploy the release mode binaries.

I could have used QuantizeValidationPass, but by the time it is invoked, min/max info is already cleared for some `CircleConst`s

File: compiler/luci/pass/src/QuantizeWeights.cpp [hunk 2]
```diff
@@ -253,6 +255,12 @@ void QuantizeWeights::quantize_weights(luci::CircleConst *weights)
       return;
     }
 
+    for (uint32_t i = 0; i < quantparam->min.size(); i++)
+      if (range_check(quantparam->min[i], quantparam->max[i], output_type))
+        WARN(l) << weights->name()
+                << " cannot be represented accurately according to chosen quantization; " << i
+                << "-th channel's range is too great" << std::endl;
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWeights.cpp:262]: Can you please open/close with curly brace for readability altough it's a single line?

Comment (seanshpark) [reply compiler/luci/pass/src/QuantizeWeights.cpp:262]: And others too

File: compiler/luci/pass/src/QuantizeBias.cpp [hunk 3]
```diff
@@ -215,6 +217,15 @@ CircleConst *QuantizeBias::quantized_bias(CircleNode *input, const CircleNode *w
       throw std::runtime_error("Unsupported quantization type.");
     }
 
+    auto qp = weight->quantparam();
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeBias.cpp:220]: I think your intention is to check qparam of bias?

File: compiler/luci/pass/src/QuantizationUtils.cpp [hunk 4]
```diff
@@ -395,4 +399,38 @@ void quant_const(luci::CircleConst *node, loco::DataType quant_type)
...
+      return 64; // a safe large default
+  }
+}
+
+} // namespace
+
+bool range_check(float min, float max, loco::DataType dtype)
+{
+  float thresh = 1.2f;
+  return log2f(max) - log2f(min) > nbits(dtype) * thresh;
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizationUtils.cpp:433]: Do you have any reference for this metric?

`min/max` is changed or ignored according to quantization schemes. So, how about using `scale` rather than `min/max`? Scale is the result of quantization algorithms, thus independent of quantization schemes. In addition, I think it would be possible to make warnings when quantizing activation.

If you'll make warnings only for const nodes, I think that quantization error (e.g., mean percentage error of `|original weights - quantized/dequantized weights|`) is a better proxy to evaluate accuracy.

File: compiler/luci/pass/src/QuantizeBias.cpp (hunk 3 above)
Comment (AShedko) [reply compiler/luci/pass/src/QuantizeBias.cpp:220]: Yes

File: compiler/luci/pass/src/QuantizationUtils.cpp (hunk 4 above)
Comment (AShedko) [reply compiler/luci/pass/src/QuantizationUtils.cpp:433]: It's what I used to find a bug with a customer model. 
`mean percentage error of |original weights - quantized/dequantized weights|` is a good metric, but it's more expensive to compute (O(N) vs O(1)).
There's a tradeoff here: how often do we want to warn the user? The more numerous warnings are more easily ignored, because they are perceived to have lower signal to noise ratio.

File: compiler/luci/pass/src/QuantizeBias.cpp (hunk 3 above)
Comment (jinevening) [reply compiler/luci/pass/src/QuantizeBias.cpp:220]: Then you'll need to update `weight` to `new_bias`

File: compiler/luci/pass/src/QuantizeBias.cpp [hunk 5]
```diff
@@ -235,6 +246,10 @@ CircleConst *QuantizeBias::quantized_bias(CircleNode *input, const CircleNode *w
     assert(weight_q->scale.size() == 1); // Only support per-layer quant
     auto weight_scale = weight_q->scale[0];
 
+    if (range_check(weight_q->min[0], weight_q->max[0], loco::DataType::S32))
+      WARN(l) << weight->name()
+              << " cannot be represented accurately according to chosen quantization" << std::endl;
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeBias.cpp:251]: This checks the range of weights, not bias.

File: compiler/luci/pass/src/QuantizationUtils.cpp (hunk 4 above)
Comment (jinevening) [reply compiler/luci/pass/src/QuantizationUtils.cpp:433]: I see. Then could you add TODO comment like,

'TODO Check if the metric is proper'

**Comment by jinevening** [comment]
@AShedko Can you move the codes to one place? Codes are located in many places, so it'll be difficult to change the metric later.

For example, make a function `warn_accuracy_with_range` and do all things in the function.

The function may be invoked right before min/max values are removed. FYI, min/max values are removed at the end of QuantizeWithMinMaxPass (in the below code).

QuantizeWithMinMax.cpp
```cpp
  // Remove min/max values
  for (auto node : loco::active_nodes(loco::output_nodes(g)))
  {
    auto circle_node = loco::must_cast<luci::CircleNode *>(node);
    if (auto qparam = circle_node->quantparam())
    {
      qparam->min.clear();
      qparam->max.clear();
    }
  }
```

File: compiler/luci/pass/src/QuantizeBias.cpp [hunk 6]
```diff
@@ -215,6 +233,7 @@ CircleConst *QuantizeBias::quantized_bias(CircleNode *input, const CircleNode *w
       throw std::runtime_error("Unsupported quantization type.");
     }
 
+    //
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeBias.cpp:236]: Unnecessary change

**Comment by AShedko** [comment]
`min` and `max` are also removed in `QuantizeWeights.cpp`:
https://github.com/Samsung/ONE/blob/0c22aacc5395190d626c5d3a3392ca4b1a881f52/compiler/luci/pass/src/QuantizeWeights.cpp#L283-L284

**Comment by jinevening** [comment]
Hmm.. how about restoring min/max from scale/zp instead of using saved values ?

For example,

For uint8,  `min = scale * (0 - zp)`, `max = scale * (255 - zp)`
For int16, `min = scale * (-32767)`, `max = scale * (32767)`

File: compiler/luci/pass/src/QuantizationUtils.cpp [hunk 7]
```diff
@@ -395,4 +395,81 @@ void quant_const(luci::CircleConst *node, loco::DataType quant_type)
...
+  auto qp = n->quantparam();
+  auto k = qp->zerop.size();
+  bool acc = false;
+  for (uint32_t i = 0; i < k; i++)
+  {
+    WARN(l) << i << "-th channel";
+    acc = acc || warn_scale_zp(qp->scale[i], qp->zerop[i], n);
+  }
+  if (acc)
+    WARN(l) << n->name() << "'s quantization may cause accuracy issues" << std::endl;
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizationUtils.cpp:472]: It would be better to move this code inside loop. E.g.,

```
for ..
{
  if (warn_scale_zp(...))
    WARN(l) "Quantization of " << i << "-th channel of " << n->name() << " may cause accuracy issues" << std::endl; 
}
```

File: compiler/luci/pass/src/QuantizationUtils.cpp [hunk 8]
```diff
@@ -395,4 +395,81 @@ void quant_const(luci::CircleConst *node, loco::DataType quant_type)
...
+// TODO Check if the metric is valid
+bool range_check(float min, float max, loco::DataType dtype)
+{
+  float thresh = 1.2f;
+  return log2f(max) - log2f(min) > nbits(dtype) * thresh;
+}
+
+bool warn_scale_zp(float scale, int64_t zp, luci::CircleNode *n)
+{
+  LOGGER(l);
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizationUtils.cpp:433]: Logging inside this function seems unnecessary.

```suggestion
```

File: compiler/luci/pass/src/QuantizationUtils.cpp [hunk 9]
```diff
@@ -395,4 +395,81 @@ void quant_const(luci::CircleConst *node, loco::DataType quant_type)
...
+      min = scale * (-32767);
+      max = scale * (32767);
+      break;
+    default:
+      return false;
+      break;
+  }
+  if (range_check(min, max, n->dtype()))
+  {
+    WARN(l) << " cannot be represented accurately according to chosen quantization" << std::endl;
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizationUtils.cpp:452]:
```suggestion
```

File: compiler/luci/pass/src/QuantizationUtils.cpp [hunk 10]
```diff
@@ -395,4 +395,81 @@ void quant_const(luci::CircleConst *node, loco::DataType quant_type)
...
+
+void warn_accuracy_with_range(luci::CircleNode *n)
+{
+  LOGGER(l);
+  auto qp = n->quantparam();
+  auto k = qp->zerop.size();
+  bool acc = false;
+  for (uint32_t i = 0; i < k; i++)
+  {
+    WARN(l) << i << "-th channel";
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizationUtils.cpp:468]:
```suggestion
```

**Comment by jinevening** [comment]
Please fix the format issue.

**Comment by AShedko** [comment]
@jinevening, PTAL?

**Review by jinevening — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM
